s2core/src/main/scala/com/kakao/s2graph/core/storage/hbase/AsynchbaseQueryBuilder.scala

```diff
@@ -91,11 +93,35 @@ class AsynchbaseQueryBuilder(storage: AsynchbaseStorage)(implicit ec: ExecutionC
                      isInnerCall: Boolean,
                      parentEdges: Seq[EdgeWithScore]): Deferred[QueryRequestWithResult] = {
 
+    def sample(edges: Seq[EdgeWithScore], n: Int): Seq[EdgeWithScore] = {
+      val pureEdges = (if (queryRequest.queryParam.offset == 0) {
+        edges.filterNot { case x => x.edge.propsPlusTs.contains(LabelMeta.degreeSeq) }
+      } else edges).toArray
+
+      val sampled = new Array[EdgeWithScore](n)
+      val N = pureEdges.size // population
+      var t = 0 // total input records dealt with
+      var m = 0 // number of items selected so far
+
+      while (m < n) {
+        val u = Random.nextDouble()
```

Comment:
@HyunsungJo  Why don't you use integer ranged RNG instead of real RNG? It much simpler and intuitive.

PS. well, I'm not familiar with Scala, though .. it does not looks like Scala-way code(like Pythonic way..) ^^;
how about refactoring this function as recursively?